### PR TITLE
refactor: migrate to `pyproject.toml`

### PR DESCRIPTION
**Pull request by Borda**
Closes #1142

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -2,73 +2,8 @@
...
-    package_data={
-        "sphinx_gallery": [
-            "_static/sg_gallery*.css",
-            "_static/no_image.png",
-            "_static/broken_example.png",
-            "_static/binder_badge_logo.svg",
-            "_static/jupyterlite_badge_logo.svg",
-        ]
-    },
     scripts=["bin/copy_sphinxgallery.sh", "bin/sphx_glr_python_to_jupyter.py"],
```
Comment (Borda) [inline setup.py:8]: TODO: need to check how to refer sh scripts...

**Comment by Borda** [comment]
@larsoner mind have a look :chipmunk:

File: pyproject.toml [hunk 2]
```diff
@@ -1,20 +1,82 @@
...
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+]
+dynamic = [
+  "version",
+]
+dependencies = [
```
Comment (larsoner) [inline pyproject.toml:32]: Looks like a little bit of unintentional information loss -- can you copy over the comment from `requirements.txt` about needing to update the README when something changes here?

File: pyproject.toml [hunk 3]
```diff
@@ -1,20 +1,82 @@
...
+[tool.setuptools]
+include-package-data = true
+
+[tool.setuptools.dynamic]
+version = {attr = "sphinx_gallery.__version__"}
+
+[tool.setuptools.packages.find]
+where = ["."]  # list of folders that contain the packages (["."] by default)
+include = ["sphinx_gallery*"]  # package names should match these glob patterns (["*"] by default)
+exclude = ["sphinx_gallery.tests*"]  # exclude packages matching these glob patterns (empty by default)
```
Comment (larsoner) [inline pyproject.toml:64]: I think we should / might as well include tests

File: MANIFEST.in [hunk 4]
```diff
@@ -2,6 +2,9 @@ include CHANGES.rst
 include LICENSE
 include ignore_words.txt
 include requirements.txt
+recursive-include sphinx_gallery *.css
+recursive-include sphinx_gallery *.png
+recursive-include sphinx_gallery *.svg
```
Comment (larsoner) [inline MANIFEST.in:7]: Again a bit of information loss I think -- the `setup.py` was more specifically scoped to `_static`, no?

**Review by larsoner — COMMENTED** [review]
Looks reasonable in principle aside from some potential differences I suspect are there. Those differences make me think that it would be safest to do a directory diff between the unzipped `sdist` and/or `.whl` on `master` and this PR to make sure we aren't adding or removing things we shouldn't.

File: pyproject.toml (hunk 3 above)
Comment (Borda) [reply pyproject.toml:64]:
```suggestion
```

File: pyproject.toml (hunk 2 above)
Comment (Borda) [inline pyproject.toml:32]:
```suggestion
# update README.rst when changing something here
dependencies = [
```

Comment (Borda) [reply pyproject.toml:32]: adding in 1b69dce87c8af4ec638ca8e5dd4ac9385794b9bb

Comment (lucyleeow) [reply pyproject.toml:32]: Is this to be added?

Comment (Borda) [reply pyproject.toml:32]: It is there already

File: MANIFEST.in (hunk 4 above)
Comment (Borda) [reply MANIFEST.in:7]: You are right, I was following the check suggestion

Comment (Borda) [reply MANIFEST.in:7]:
```suggestion
include sphinx_gallery/_static/sg_gallery*.css
include sphinx_gallery/ _static/no_image.png
include sphinx_gallery/_static/broken_example.png
include sphinx_gallery/_static/binder_badge_logo.svg
include sphinx_gallery/_static/jupyterlite_badge_logo.svg
```

File: MANIFEST.in [hunk 5]
```diff
@@ -2,6 +2,11 @@ include CHANGES.rst
 include LICENSE
 include ignore_words.txt
 include requirements.txt
+include sphinx_gallery/_static/sg_gallery*.css
+include sphinx_gallery/ _static/no_image.png
```
Comment (larsoner) [inline MANIFEST.in:6]:
```suggestion
include sphinx_gallery/_static/no_image.png
```

**Comment by Borda** [comment]
> Can you try the directory diff and let me know how it looks?

not sure what you mean, like build a package from master and with this suggestion and see if there is a difference in the package?

**Comment by Borda** [comment]
this is the version from master:
```
dist/master/
└── sphinx-gallery-0.16.dev0
    ├── CHANGES.rst
    ├── examples
    │   ├── local_module.py
    │   ├── no_output
    │   │   ├── just_code.py
    │   │   ├── plot_raise.py
    │   │   ├── plot_strings.py
    │   │   ├── plot_syntaxerror.py
    │   │   └── README.txt
    │   ├── plot_0_sin.py
    │   ├── plot_1_exp.py
    │   ├── plot_2_seaborn.py
    │   ├── plot_3_capture_repr.py
    │   ├── plot_4b_provide_thumbnail.py
    │   ├── plot_4_choose_thumbnail.py
    │   ├── plot_5_unicode_everywhere.py
    │   ├── plot_6_function_identifier.py
    │   ├── plot_7_sys_argv.py
    │   ├── plot_8_animations.py
    │   └── README.txt
    ├── ignore_words.txt
    ├── LICENSE
    ├── MANIFEST.in
    ├── PKG-INFO
    ├── plotly_examples
    │   ├── plot_0_plotly.py
    │   └── README.rst
    ├── pyproject.toml
    ├── pyvista_examples
    │   ├── plot_collisions.py
    │   ├── plot_glyphs.py
    │   ├── plot_lighting.py
    │   ├── plot_ray_trace.py
    │   └── README.rst
    ├── README.rst
    ├── requirements.txt
    ├── setup.cfg
    ├── setup.py
    ├── sphinx_gallery
    │   ├── backreferences.py
    │   ├── block_parser.py
    │   ├── directives.py
    │   ├── docs_resolv.py
    │   ├── downloads.py
    │   ├── gen_gallery.py
    │   ├── gen_rst.py
    │   ├── __init__.py
    │   ├── interactive_example.py
    │   ├── load_style.py
    │   ├── notebook.py
    │   ├── py_source_parser.py
    │   ├── recommender.py
    │   ├── scrapers.py
    │   ├── sorting.py
    │   ├── _static
    │   │   ├── binder_badge_logo.svg
    │   │   ├── broken_example.png
    │   │   ├── jupyterlite_badge_logo.svg
    │   │   ├── no_image.png
    │   │   ├── sg_gallery-binder.css
    │   │   ├── sg_gallery.css
    │   │   ├── sg_gallery-dataframe.css
    │   │   └── sg_gallery-rendered-html.css
    │   ├── tests
    │   │   ├── conftest.py
    │   │   ├── __init__.py
    │   │   ├── reference_parse.txt
    │   │   ├── test_backreferences.py
    │   │   ├── test_block_parser.py
    │   │   ├── testconfs
    │   │   │   ├── index.rst
    │   │   │   └── src
    │   │   │       ├── plot_1.py
    │   │   │       ├── plot_2.py
    │   │   │       ├── plot_3.py
    │   │   │       └── README.txt
    │   │   ├── test_docs_resolv.py
    │   │   ├── test_full_noexec.py
    │   │   ├── test_full.py
    │   │   ├── test_gen_gallery.py
    │   │   ├── test_gen_rst.py
    │   │   ├── test_interactive_example.py
    │   │   ├── test_load_style.py
    │   │   ├── test_notebook.py
    │   │   ├── test_py_source_parser.py
    │   │   ├── test_recommender.py
    │   │   ├── test_scrapers.py
    │   │   ├── test_sorting.py
    │   │   └── tinybuild
    │   │       ├── doc
    │   │       │   ├── binder
    │   │       │   │   └── requirements.txt
    │   │       │   ├── conf.py
    │   │       │   ├── index.rst
    │   │       │   ├── make.bat
    │   │       │   ├── Makefile
    │   │       │   ├── minigallery.rst
    │   │       │   ├── _static_nonstandard
    │   │       │   │   └── demo.png
    │   │       │   └── _templates
    │   │       │       └── module.rst
    │   │       ├── examples
    │   │       │   ├── future
    │   │       │   │   ├── plot_future_imports_broken.py
    │   │       │   │   ├── plot_future_imports.py
    │   │       │   │   └── README.rst
    │   │       │   ├── __init__.py
    │   │       │   ├── isentropic.m
    │   │       │   ├── julia_sample.jl
    │   │       │   ├── local_module.py
    │   │       │   ├── parse_this.cpp
    │   │       │   ├── plot_animation.py
    │   │       │   ├── plot_command_line_args.py
    │   │       │   ├── plot_defer_figures.py
    │   │       │   ├── plot_log.py
    │   │       │   ├── plot_matplotlib_alt.py
    │   │       │   ├── plot_matplotlib_backend.py
    │   │       │   ├── plot_numpy_matplotlib.py
    │   │       │   ├── plot_pickle.py
    │   │       │   ├── plot_repr.py
    │   │       │   ├── plot_scraper_broken.py
    │   │       │   ├── plot_second_future_imports.py
    │   │       │   ├── plot_svg.py
    │   │       │   ├── plot_webp.py
    │   │       │   └── README.txt
    │   │       ├── examples_rst_index
    │   │       │   ├── examp_subdir1
    │   │       │   │   ├── plot_sub1.py
    │   │       │   │   └── README.txt
    │   │       │   ├── examp_subdir2
    │   │       │   │   ├── index.rst
    │   │       │   │   └── plot_sub2.py
    │   │       │   ├── index.rst
    │   │       │   └── plot_examp.py
    │   │       └── examples_with_rst
    │   │           ├── plot_boo.py
    │   │           ├── plot_cos.py
    │   │           ├── README.txt
    │   │           ├── rst_example1.rst
    │   │           └── rst_example2.rst
    │   └── utils.py
    ├── sphinx_gallery.egg-info
    │   ├── dependency_links.txt
    │   ├── entry_points.txt
    │   ├── PKG-INFO
    │   ├── requires.txt
    │   ├── SOURCES.txt
    │   └── top_level.txt
    └── tutorials
        ├── plot_parse.py
        └── README.txt
```
and this is with this PR:
```
dist/pr-pyproject/
└── sphinx-gallery-0.16.dev0
    ├── CHANGES.rst
    ├── examples
    │   ├── local_module.py
    │   ├── no_output
    │   │   ├── just_code.py
    │   │   ├── plot_raise.py
    │   │   ├── plot_strings.py
    │   │   ├── plot_syntaxerror.py
    │   │   └── README.txt
    │   ├── plot_0_sin.py
    │   ├── plot_1_exp.py
    │   ├── plot_2_seaborn.py
    │   ├── plot_3_capture_repr.py
    │   ├── plot_4b_provide_thumbnail.py
    │   ├── plot_4_choose_thumbnail.py
    │   ├── plot_5_unicode_everywhere.py
    │   ├── plot_6_function_identifier.py
    │   ├── plot_7_sys_argv.py
    │   ├── plot_8_animations.py
    │   └── README.txt
    ├── ignore_words.txt
    ├── LICENSE
    ├── MANIFEST.in
    ├── PKG-INFO
    ├── plotly_examples
    │   ├── plot_0_plotly.py
    │   └── README.rst
    ├── pyproject.toml
    ├── pyvista_examples
    │   ├── plot_collisions.py
    │   ├── plot_glyphs.py
    │   ├── plot_lighting.py
    │   ├── plot_ray_trace.py
    │   └── README.rst
    ├── README.rst
    ├── setup.cfg
    ├── sphinx_gallery
    │   ├── backreferences.py
    │   ├── block_parser.py
    │   ├── directives.py
    │   ├── docs_resolv.py
    │   ├── downloads.py
    │   ├── gen_gallery.py
    │   ├── gen_rst.py
    │   ├── __init__.py
    │   ├── interactive_example.py
    │   ├── load_style.py
    │   ├── notebook.py
    │   ├── py_source_parser.py
    │   ├── recommender.py
    │   ├── scrapers.py
    │   ├── sorting.py
    │   ├── _static
    │   │   ├── binder_badge_logo.svg
    │   │   ├── broken_example.png
    │   │   ├── jupyterlite_badge_logo.svg
    │   │   ├── no_image.png
    │   │   ├── sg_gallery-binder.css
    │   │   ├── sg_gallery.css
    │   │   ├── sg_gallery-dataframe.css
    │   │   └── sg_gallery-rendered-html.css
    │   ├── tests
    │   │   ├── conftest.py
    │   │   ├── __init__.py
    │   │   ├── reference_parse.txt
    │   │   ├── test_backreferences.py
    │   │   ├── test_block_parser.py
    │   │   ├── testconfs
    │   │   │   ├── index.rst
    │   │   │   └── src
    │   │   │       ├── plot_1.py
    │   │   │       ├── plot_2.py
    │   │   │       ├── plot_3.py
    │   │   │       └── README.txt
    │   │   ├── test_docs_resolv.py
    │   │   ├── test_full_noexec.py
    │   │   ├── test_full.py
    │   │   ├── test_gen_gallery.py
    │   │   ├── test_gen_rst.py
    │   │   ├── test_interactive_example.py
    │   │   ├── test_load_style.py
    │   │   ├── test_notebook.py
    │   │   ├── test_py_source_parser.py
    │   │   ├── test_recommender.py
    │   │   ├── test_scrapers.py
    │   │   ├── test_sorting.py
    │   │   └── tinybuild
    │   │       ├── doc
    │   │       │   ├── binder
    │   │       │   │   └── requirements.txt
    │   │       │   ├── conf.py
    │   │       │   ├── index.rst
    │   │       │   ├── make.bat
    │   │       │   ├── Makefile
    │   │       │   ├── minigallery.rst
    │   │       │   ├── _static_nonstandard
    │   │       │   │   └── demo.png
    │   │       │   └── _templates
    │   │       │       └── module.rst
    │   │       ├── examples
    │   │       │   ├── future
    │   │       │   │   ├── plot_future_imports_broken.py
    │   │       │   │   ├── plot_future_imports.py
    │   │       │   │   └── README.rst
    │   │       │   ├── __init__.py
    │   │       │   ├── isentropic.m
    │   │       │   ├── julia_sample.jl
    │   │       │   ├── local_module.py
    │   │       │   ├── parse_this.cpp
    │   │       │   ├── plot_animation.py
    │   │       │   ├── plot_command_line_args.py
    │   │       │   ├── plot_defer_figures.py
    │   │       │   ├── plot_log.py
    │   │       │   ├── plot_matplotlib_alt.py
    │   │       │   ├── plot_matplotlib_backend.py
    │   │       │   ├── plot_numpy_matplotlib.py
    │   │       │   ├── plot_pickle.py
    │   │       │   ├── plot_repr.py
    │   │       │   ├── plot_scraper_broken.py
    │   │       │   ├── plot_second_future_imports.py
    │   │       │   ├── plot_svg.py
    │   │       │   ├── plot_webp.py
    │   │       │   └── README.txt
    │   │       ├── examples_rst_index
    │   │       │   ├── examp_subdir1
    │   │       │   │   ├── plot_sub1.py
    │   │       │   │   └── README.txt
    │   │       │   ├── examp_subdir2
    │   │       │   │   ├── index.rst
    │   │       │   │   └── plot_sub2.py
    │   │       │   ├── index.rst
    │   │       │   └── plot_examp.py
    │   │       └── examples_with_rst
    │   │           ├── plot_boo.py
    │   │           ├── plot_cos.py
    │   │           ├── README.txt
    │   │           ├── rst_example1.rst
    │   │           └── rst_example2.rst
    │   └── utils.py
    ├── sphinx_gallery.egg-info
    │   ├── dependency_links.txt
    │   ├── entry_points.txt
    │   ├── PKG-INFO
    │   ├── requires.txt
    │   ├── SOURCES.txt
    │   └── top_level.txt
    └── tutorials
        ├── plot_parse.py
        └── README.txt
```
the only diff is as expected missing:
```
    ├── requirements.txt
	...
    ├── setup.py
```

**Comment by larsoner** [comment]
Thanks @Borda !

**Comment by lucyleeow** [comment]
Thank you @Borda :pray: !

**Pull request merged**